qiskit_machine_learning/algorithms/classifiers/neural_network_classifier.py

```diff
@@ -119,4 +124,34 @@ def predict(self, X: np.ndarray) -> np.ndarray:  # pylint: disable=invalid-name
     def score(
         self, X: np.ndarray, y: np.ndarray, sample_weight: Optional[np.ndarray] = None
     ) -> float:
+        y = self._convert_categorical(y)
         return ClassifierMixin.score(self, X, y, sample_weight)
+
+    def _convert_categorical(self, y: np.ndarray) -> np.ndarray:
+        """Converts categorical target data using label or one-hot encoding."""
+        y = np.array(y)
+
+        # Test if data is numeric
+        try:
+            y.astype(float)
+        except ValueError:
+            # non-numeric data is assumed to be categorical
+            is_categorical = True
+            y = y.reshape(-1, 1)
+        else:
+            is_categorical = False
+
+        if is_categorical and self._one_hot:
+            if not self._target_encoder:
+                self._target_encoder = OneHotEncoder()
+                self._target_encoder.fit(y)
+            y = self._target_encoder.transform(y)
+        elif is_categorical and not self._one_hot:
+            if not self._target_encoder:
+                self._target_encoder = LabelEncoder()
+                self._target_encoder.fit(y)
+            y = self._target_encoder.transform(y)
```

Comment:
I think this should be more like `if self._target_encoder is None:`. However I wonder so you need to store this in an instance var? I only see it being used here. How about

```
        if is_categorical:
            target_encoder = OneHotEncoder() if self._one_hot else LabelEncoder()
            target_encoder.fit(y)
            y = target_encoder.transform(y)
```        
which keeps target encoder local, and condenses the code since its really the same in each branch except for the encoder.    (I was looking at this since mypy failed in this area)

Reply:
FYI. CI is now running again. There is some random failure that sometimes happen, that we should take care of due to precision, but is unrelated to your contribution. Other jobs are failing mypy as I mentioned related to the above code
```
qiskit_machine_learning/algorithms/classifiers/neural_network_classifier.py:147: error: "None" has no attribute "fit"
qiskit_machine_learning/algorithms/classifiers/neural_network_classifier.py:148: error: "None" has no attribute "transform"
qiskit_machine_learning/algorithms/classifiers/neural_network_classifier.py:152: error: "None" has no attribute "fit"
qiskit_machine_learning/algorithms/classifiers/neural_network_classifier.py:153: error: "None" has no attribute "transform"
```

Reply:
@woodsp-ibm the only reason I store the encoder as an instance var is becasue of the potential edge case where a user calls `score()` on categorical test data that is missing data for a certain class. For example, if they train on data that has the classes "A", "B", and "C" but then call score and provide it a subset of data that happens to only have "A" and "C" - this would result in an inconsistent encoding. I don't think this would be very common becuase the test data should be balanced but it could still happen. 
Can you think of a better way to enforce the output of the encoding? Maybe I could try using the `output_shape` attribute of the `NeuralNetwork` or requiring that the user provide a list of classes to fit the encoder to? Should I even be concerned with this edgecase?

I agree with your other comments though and will make some changes. 
Thanks for taking care of the spelling mistakes btw!

Reply:
So the encoder is stateful I guess which is why you need to re-use it for score. Since the type really only depends on the one_hot setting would it be simpler then to create it in the constructor, rather than doing it lazily later. That would save the test of it being there or not and the code would be simply something like
```
        if is_categorical:
            self._target_encoder.fit(y)
            y = self._target_encoder.transform(y)
``` 
I am not so familiar with this area; if I then train later with different data say with more classes can I reuse the same encoder - in your logic once its created it will stay there and be re-used. On that basis is why I was suggesting creating in constructor - just checking (or do you need to create a new one when fitting each time, which I don;t see the code doing)

Reply:
Ya the reason why I don't instantiate the proper encoder in the constructor and use `None` instead is because I wanted ensure that we call `encoder.fit()` only once to ensure the encoding stays consistent between all calls to the classifier's `score` and `fit` method. The other way I can think to ensure it is fit only once would be to call `encoder.transform()` in a try-catch block.